src/realm/object-store/impl/deep_change_checker.cpp

```diff
@@ -226,25 +289,201 @@ bool DeepChangeChecker::check_row(Table const& table, ObjKeyType key, size_t dep
+
+        // For the special case of having a backlink at the end of a key path we need to check this level too.
+        // Modifications to a backlink are found via the insertions on the origin table (which we are in right
+        // now).
+        auto last_key_path_element = key_path[key_path.size() - 1];
+        auto last_column_key = last_key_path_element.second;
+        if (last_column_key.get_type() == col_type_BackLink) {
+            auto iterator = m_info.tables.find(table.get_key().value);
+            if (iterator != m_info.tables.end() && !iterator->second.insertions_empty()) {
+                auto root_column_key = key_path[0].second;
+                changed_columns.push_back(root_column_key.value);
+            }
+        }
+
+        return;
+    }
+
+    auto [table_key, column_key] = key_path.at(depth);
+
+    // Check for a change on the current depth level.
+    auto iterator = m_info.tables.find(table_key.value);
+    if (iterator != m_info.tables.end() && (iterator->second.modifications_contains(object_key_value, {column_key}) ||
+                                            iterator->second.insertions_contains(object_key_value))) {
+        // If an object linked to the root object was changed we only mark the
+        // property of the root objects as changed.
+        // This is also the reason why we can return right after doing so because we would only mark the same root
+        // property again in case we find another change deeper down the same path.
+        auto root_column_key = key_path[0].second;
+        changed_columns.push_back(root_column_key.value);
+        return;
+    }
+
+    // Only continue for any kind of link.
+    auto column_type = column_key.get_type();
+    if (column_type != col_type_Link && column_type != col_type_LinkList && column_type != col_type_BackLink &&
+        column_type != col_type_TypedLink && column_type != col_type_Mixed) {
+        return;
+    }
+
+    auto check_mixed_object = [&](const Mixed& mixed_object) {
+        if (mixed_object.is_type(type_Link, type_TypedLink)) {
+            auto object_key = mixed_object.get<ObjKey>();
+            auto target_table_key = mixed_object.get_link().get_table_key();
+            Group* group = table.get_parent_group();
+            auto target_table = group->get_table(target_table_key);
+            find_changed_columns(changed_columns, key_path, depth + 1, *target_table, object_key.value);
+        }
+    };
+
+    // Advance one level deeper into the key path.
+    auto object = table.get_object(ObjKey(object_key_value));
+    auto target_table = table.get_link_target(column_key);
+    if (column_key.is_list()) {
+        if (column_type == col_type_Mixed) {
+            auto list = object.get_list<Mixed>(column_key);
+            for (size_t i = 0; i < list.size(); i++) {
+                auto target_object = list.get_any(i);
+                check_mixed_object(target_object);
+            }
+        }
+        else {
```

Comment:
all these else blocks are a good place to add some assertions that the column there is really the correct type that you are expecting, this can help us find errors especially if we add new link types in the future (this is not a blocking comment but just a general note about adding more assertions which can help us debug things in the future)

Reply:
Added with 8812cdd25.